release/apparmor.go

```diff
@@ -97,117 +184,100 @@ var (
 		"ptrace",
 		"signal",
 	}
+	// Since AppArmorParserMtime() will be called by generateKey() in
+	// system-key and that could be called by different users on the
+	// system, use a predictable search path for finding the parser.
+	appArmorParserSearchPath = "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin"
+	// Each apparmor feature is manifested as a directory entry.
+	appArmorFeaturesSysPath = "/sys/kernel/security/apparmor/features"
 )
 
-// isDirectoy is like osutil.IsDirectory but we cannot import this
-// because of import cycles
-func isDirectory(path string) bool {
-	stat, err := os.Stat(path)
-	if err != nil {
-		return false
+func assessAppArmor() {
+	kernelFeatures := AppArmorKernelFeatures()
+
+	if len(kernelFeatures) == 0 {
+		appArmorLevel = NoAppArmor
+		appArmorSummary = "apparmor not enabled"
+		return
 	}
-	return stat.IsDir()
-}
 
-func probeAppArmor() (AppArmorLevelType, string) {
-	if !isDirectory(appArmorFeaturesSysPath) {
-		return NoAppArmor, "apparmor not enabled"
+	parserFeatures := AppArmorParserFeatures()
+	var missingParserFeatures []string
+	for _, feature := range requiredAppArmorParserFeatures {
+		if !strutil.SortedListContains(parserFeatures, feature) {
```

Comment:
> IME we should return UnusableAppArmor if `len(parserFeatures) == 0` since AppArmorParserFeatures() no longer returns an error but instead an empty list when the parser is not found. As it stands, this PR is operationally correct but there is an 'almost bug' here in that if we ever decide to remove 'unsafe' as a required feature then there is no way to discern between a missing parser and no required features.

Thanks for all the updates! This is the only remaining item for my +1

Reply:
+1

Reply:
What is the idea? How involved is it?

Reply:
To have two sets: features that presence of yields full level and features that absence of yields unsupported level. It's not too involved I suspect.

Reply:
Well, the current code has a concept of missing parser error vs empty features so removing that to only add it back in a separate PR seems unnecessary (especially since it should be very complicated). Am I missing something?

Reply:
I believe this is ready now.